internal/ddltimer/ddltimer.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package ddltimer includes a [DeadlineTimer] that can be used to set deadlines and listen for time-out events. Here is
+an example of how to use the DeadlineTimer:
+
+	t := ddltimer.New()
+	defer t.Stop()  // to prevent resource leaks
+	t.SetDeadline(time.Now().Add(2 * time.Second))
+	<-t.Timeout()   // will return after 2 seconds
+	// you may also SetDeadline in other goroutines while waiting
+*/
+package ddltimer
+
+import (
+	"sync"
+	"time"
+)
+
+// DeadlineTimer is a tool that allows you to set deadlines and listen for time-out events. It is more flexible than
+// [time.After] because you can update the deadline; and it is more flexible than [time.Timer] because multiple
+// subscribers can listen to the time-out channel.
+//
+// DeadlineTimer is safe for concurrent use by multiple goroutines.
+//
+// gvisor has a similar implementation: [gonet.deadlineTimer].
+//
+// [gonet.deadlineTimer]: https://github.com/google/gvisor/blob/release-20230605.0/pkg/tcpip/adapters/gonet/gonet.go#L130-L138
+type DeadlineTimer struct {
+	mu sync.Mutex
+
+	ddl time.Time
+	t   *time.Timer
+	c   chan struct{}
+}
+
+// New creates a new instance of DeadlineTimer that can be used to SetDeadline() and listen for Timeout() events.
+func New() *DeadlineTimer {
+	return &DeadlineTimer{
+		c: make(chan struct{}),
+	}
+}
+
+// Timeout returns a readonly channel that will block until the specified amount of time set by SetDeadline() has
+// passed. This channel can be safely subscribed to by multiple listeners.
+//
+// Timeout is similar to the [time.After] function.
+func (d *DeadlineTimer) Timeout() <-chan struct{} {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	return d.c
+}
+
+// SetDeadline changes the timer to expire after deadline t. When the timer expires, the Timeout() channel will be
+// unblocked. A zero value means the timer will not time out.
+//
+// SetDeadline is like [time.Timer]'s Reset() function, but it doesn't have any restrictions.
+func (d *DeadlineTimer) SetDeadline(t time.Time) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Stop the timer, and if d.t has already invoked the callback of AfterFunc, create a new channel.
+	if d.t != nil && !d.t.Stop() {
```

Comment:
I'll point out that the timer we need is for a one-time expiration (unlike net.Conn's). We don't need to handle resets after the initial expiration, which simplifies things.

Reply:
Yes, this is valid argument. I can try `AfterFunc` and `Reset` to see whether it works. But `NewTimer` and `Reset` is very hard to be used in multi-thread environment.

Reply:
updated to use `time.AfterFunc`.